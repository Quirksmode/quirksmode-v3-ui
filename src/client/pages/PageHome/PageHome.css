.PageHome {
  &__colWrap {
    @media (--tablet-portrait) {
      display: flex;
      justify-content: space-between;
    }
  }

  &__col {
    @media (--tablet-portrait) {
      margin-right: 5.6%;
      width: 50%;
    }

    &:last-child {
      margin-right: 0;
    }
  }
}
