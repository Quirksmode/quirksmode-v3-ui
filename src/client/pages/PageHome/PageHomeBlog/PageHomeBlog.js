import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import PostItem from 'components/PostItem/PostItem';

/**
 * Description
 *
 * @name PageHomeBlog
 * @param  {object} props.latestBlogPosts []
 */
const PageHomeBlog = ({
  latestBlogPosts,
  action
}) => (
  <div className="PageHomeBlog">
    <div className="Page__headingWrap">
      <h2>General Shoutings</h2>
      <NavLink
        to="/blog"
        className="link--withHeading link--withIcon"
      >
        View My Blog
      </NavLink>
    </div>
    <div className="PageHomeBlog__posts">
      { latestBlogPosts.map(post => (
        <PostItem
          key={ post.postID }
          post={ post }
          action={ action }
          size="Small"
          lineClamp="3"
        />
      ))}
    </div>
  </div>
);

PageHomeBlog.propTypes = {
  latestBlogPosts: PropTypes.array,
  action: PropTypes.func
};

export default PageHomeBlog;
