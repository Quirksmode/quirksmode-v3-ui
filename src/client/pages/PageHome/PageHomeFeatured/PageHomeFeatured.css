.PageHomeFeatured {
  .grid {
    max-width: 1012px;
  }

  .Page__headingWrap {
    padding: 0 6px;
  }

  .swiper-container {
    padding: 6px;
  }

  .swiper-wrapper {
    display: flex;

    /* Show a grid when JS is turned off */
    .no-js & {
      display: grid;
      grid-gap: 20px;
      grid-template-columns: 1fr;

      @media (--custom-370) {
        grid-gap: 30px;
        grid-template-columns: 1fr 1fr;
      }

      @media (--tablet-portrait) {
        grid-template-columns: 1fr 1fr 1fr;
      }

      @media (--custom-1000) {
        grid-gap: 56px;
      }
    }
  }

  /* Prevent UI jump when initializing */
  &:not(.swiper-container-initialized) {
    .swiper-slide {
      .js & {
        margin-right: 5%;
        width: 30%;
      }
    }
  }

  .Slider {
    &__directionNav {
      display: none;
      opacity: 0;

      @media (--mobile-landscape) {
        display: block;
      }

      &--active {
        opacity: 1;
      }
    }
  }
}
