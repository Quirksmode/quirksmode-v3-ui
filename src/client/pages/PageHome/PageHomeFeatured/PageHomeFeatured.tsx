import React, {
  useState
} from 'react';
import PropTypes from 'prop-types';
import Swiper from 'react-id-swiper';
import { useInView } from 'react-intersection-observer';
import { NavLink } from 'react-router-dom';
import Frame from 'components/Frame/Frame';
import LinkLoader from 'components/LinkLoader/LinkLoader';

/**
 * Home Page Featured Component
 *
 * @name PageHomeSkills
 * @param {array} props.featuredWork - Featured Projects
 * @return {JSXElement}
 */
const PageHomeFeatured = ({
  featuredWork
}) => {
  /**
   * React State mutator for setting the swiper instance
   *
   * @type {[object, function]}
   */
  const [swiper, updateSwiper] = useState(null);

  /**
   * Params for the Swiper component
   *
   * @type {Object}
   */
  const params = {
    preloadImages: false,
    lazyLoading: true,
    updateOnImagesReady: true,
    slidesPerView: 1,
    spaceBetween: 10,
    loop: false,
    autoHeight: true,
    breakpoints: {
      481: {
        slidesPerView: 2,
        spaceBetween: 30
      },
      768: {
        slidesPerView: 3,
        spaceBetween: 50
      },
    }
  };

  /**
   * Go to the next slide
   *
   * @name goNext
   * @type {function}
   */
  const goNext = () => {
    if (swiper !== null) {
      swiper.slideNext();
    }
  };

  /**
   * Go to the previous slide
   *
   * @name goPrev
   * @type {function}
   */
  const goPrev = () => {
    if (swiper !== null) {
      swiper.slidePrev();
    }
  };

  /**
   * Hook that is generated by the useInView function from the
   * react-intersection-observer library.
   *
   * @type {[object, function]}
   */
  const [sliderNavRef, inView] = useInView({
    threshold: 1,
    rootMargin: '-150px'
  });

  return (
    <section className="PageHomeFeatured Page__section Page__section--greyFade">
      <div className="Page__sectionInner grid">
        <div className="Page__headingWrap">
          <h2>Featured Work</h2>
          <NavLink
            to="/portfolio"
            className="link--withHeading link--withIcon"
          >
            View All Work
          </NavLink>
        </div>
        <Swiper
          getSwiper={ updateSwiper }
          { ...params }
        >
          { featuredWork.map((project) => {
            const {
              id,
              slug,
              featuredImage,
              title,
              isNew
            } = project;
            return (
              <div className="PageHomeFeatured__slide" key={ id }>
                <LinkLoader
                  type="portfolio"
                  slug={ slug }
                >
                  <Frame
                    featuredImage={ featuredImage }
                    title={ title }
                    isNew={ isNew }
                    size="Medium"
                  />
                </LinkLoader>
              </div>
            );
          })}
        </Swiper>
      </div>
      <div
        ref={ sliderNavRef }
        className={ `Slider__directionNav${inView ? ' Slider__directionNav--active' : ''}` }
      >
        <button
          type="button"
          onClick={ goPrev }
          className="Slider__prev"
        >
          Previous
        </button>
        <button
          type="button"
          onClick={ goNext }
          className="Slider__next"
        >
          Next
        </button>
      </div>
    </section>
  );
};

PageHomeFeatured.propTypes = {
  featuredWork: PropTypes.array
};

export default PageHomeFeatured;
