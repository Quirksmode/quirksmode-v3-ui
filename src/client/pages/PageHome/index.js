import React from 'react';
import loadable from '@loadable/component';
import Loading from 'components/Loading/Loading';
import ErrorBoundary from 'components/ErrorBoundary/ErrorBoundary';

const Home = loadable(() => import('./PageHome'), {
  fallback: <Loading />
});

const WrappedHome = props => (
  <ErrorBoundary>
    <Home { ...props } />
  </ErrorBoundary>
);

export default WrappedHome;
