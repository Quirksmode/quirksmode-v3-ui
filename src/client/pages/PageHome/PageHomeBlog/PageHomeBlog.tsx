import React from 'react';
import { NavLink } from 'react-router-dom';
import PostItem from 'components/PostItem/PostItem';
import { PageHomeBlogProps } from './PageHomeBlog.types';

/**
 * Home Page Latest Blog Posts
 */
const PageHomeBlog: React.FC<PageHomeBlogProps> = ({ latestBlogPosts }) => (
  <div className="PageHomeBlog">
    <div className="Page__headingWrap">
      <h2>General Shoutings</h2>
      <NavLink to="/blog" className="link--withHeading link--withIcon">
        View My Blog
      </NavLink>
    </div>
    <div className="PageHomeBlog__posts">
      {latestBlogPosts.map((post) => (
        <PostItem key={post.postID} post={post} size="Small" lineClamp="3" />
      ))}
    </div>
  </div>
);

export default PageHomeBlog;
