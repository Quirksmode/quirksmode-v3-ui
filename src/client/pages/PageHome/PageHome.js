import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Meta from 'components/Meta/Meta';
import PageWrapper from 'components/PageWrapper/PageWrapper';
import { fetchHomeData } from './PageHome.actions';
import PageHomeFeatured from './PageHomeFeatured/PageHomeFeatured';
import PageHomeIntro from './PageHomeIntro/PageHomeIntro';
import PageHomeSkills from './PageHomeSkills/PageHomeSkills';
import PageHomeBlog from './PageHomeBlog/PageHomeBlog';

/**
 * Home Page
 *
 * @name PageHome
 * @param {function} props.fetchHomeDataAction - Redux action to Fetch the Home Page Data
 * @param {object} props.content - The content for this Page
 * @param {object} props.metadata - The metadata for this Page
 * @param {boolean} props.loading - Flag for while the data is being fetched
 * @param {boolean} props.error - Flag for if there is an error fetching the data
 * @param {object} props.siteSettings - The Global Site Settings
 * @return {JSXElement}
 */
const PageHome = ({
  fetchHomeDataAction,
  content,
  metadata,
  loading,
  error,
  siteSettings
}) => {
  /**
   * @type {Object}
   * @property {array} content.featuredWork - For the featured work Slider
   * @property {object} content.intro - Intro content
   * @property {array} content.latestBlogPosts - Latest Blog Posts
   * @property {array} content.skills - List of skills
   */
  const {
    featuredWork,
    intro,
    latestBlogPosts,
    skills
  } = content;

  /**
   * @type {Object}
   * @property {object} siteSettings.cv - CV data from the Global Site Settings
   */
  const {
    cv
  } = siteSettings;

  /**
   * Fetch the Home Page Data via Redux, but only if the data does not already exist
   */
  useEffect(() => {
    if (!intro) fetchHomeDataAction();
  }, [fetchHomeDataAction, intro]);

  return (
    <PageWrapper error={ error } loading={ loading }>
      <div className="page PageHome">
        <Meta { ...metadata } />
        <h1 className="visuallyHidden">Welcome to Quirksmode</h1>
        { featuredWork.length > 0 && <PageHomeFeatured featuredWork={ featuredWork } /> }
        <section className="Page__section">
          <div className="Page__sectionInner grid">
            <div className="PageHome__colWrap">
              <div className="PageHome__col">
                { intro && <PageHomeIntro intro={ intro } cv={ cv } /> }
                { skills.length > 0 && <PageHomeSkills skills={ skills } /> }
              </div>
              <div className="PageHome__col">
                <PageHomeBlog
                  latestBlogPosts={ latestBlogPosts }
                  loading={ loading }
                />
              </div>
            </div>
          </div>
        </section>
      </div>
    </PageWrapper>
  );
};

PageHome.propTypes = {
  fetchHomeDataAction: PropTypes.func,
  content: PropTypes.object,
  metadata: PropTypes.object,
  loading: PropTypes.bool,
  error: PropTypes.bool,
  siteSettings: PropTypes.object
};

const mapStateToProps = ({
  pageHome,
  app
}) => ({
  content: pageHome.content,
  metadata: pageHome.metadata,
  loading: pageHome.loading,
  error: pageHome.error,
  siteSettings: app.siteSettings
});

const mapDispatchToProps = dispatch => ({
  fetchHomeDataAction: (...args) => dispatch(fetchHomeData(...args))
});

export default connect(mapStateToProps, mapDispatchToProps)(PageHome);
