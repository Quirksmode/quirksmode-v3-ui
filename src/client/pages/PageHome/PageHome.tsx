import React, { useEffect, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import { useTypedSelector } from 'client/redux/types';
import Meta from 'components/Meta/Meta';
import PageWrapper from 'components/PageWrapper/PageWrapper';
import { fetchHomeData } from './PageHome.actions';
import PageHomeFeatured from './PageHomeFeatured/PageHomeFeatured';
import PageHomeIntro from './PageHomeIntro/PageHomeIntro';
import PageHomeSkills from './PageHomeSkills/PageHomeSkills';
import PageHomeBlog from './PageHomeBlog/PageHomeBlog';

/**
 * Home Page
 */
const PageHome: React.FC = () => {
  const dispatch = useDispatch();
  const pageHome = useTypedSelector((state) => state.pageHome);
  const app = useTypedSelector((state) => state.app);

  const { content, metadata, loading, error } = pageHome;
  const { featuredWork, intro, latestBlogPosts, skills } = content;
  const { cv } = app.siteSettings;

  /**
   * Fetch the Home Page Data via Redux, but only if the data does not already exist
   */
  useEffect(() => {
    if (!intro) {
      dispatch(fetchHomeData());
    }
  }, [intro]);

  return useMemo(
    () => (
      <PageWrapper error={error} loading={loading}>
        <div className="page PageHome">
          {metadata && <Meta {...metadata} />}
          <h1 className="visuallyHidden">Welcome to Quirksmode</h1>
          {featuredWork?.length > 0 && (
            <PageHomeFeatured featuredWork={featuredWork} />
          )}
          <section className="Page__section">
            <div className="Page__sectionInner grid">
              <div className="PageHome__colWrap">
                <div className="PageHome__col">
                  {intro && <PageHomeIntro intro={intro} cv={cv} />}
                  {skills?.length > 0 && <PageHomeSkills skills={skills} />}
                </div>
                <div className="PageHome__col">
                  {latestBlogPosts?.length > 0 && (
                    <PageHomeBlog latestBlogPosts={latestBlogPosts} />
                  )}
                </div>
              </div>
            </div>
          </section>
        </div>
      </PageWrapper>
    ),
    [app, content, error, loading, metadata]
  );
};

export default PageHome;
