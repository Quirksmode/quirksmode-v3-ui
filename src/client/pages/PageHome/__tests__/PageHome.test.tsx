import { mountRender, reduxHelper } from 'client/tests/testUtils';
import moxios from 'moxios';
import PageHome from '../PageHome';
import AppStub from 'client/__tests__/app.stub.json';
import PageHomeStub from './PageHome.stub.json';
import 'tests/mocks/intersectionObserver';
import PageHomeBlog from '../PageHomeBlog/PageHomeBlog';
import PageWrapper from 'client/components/PageWrapper/PageWrapper';
import PageHomeFeatured from '../PageHomeFeatured/PageHomeFeatured';
import PageHomeIntro from '../PageHomeIntro/PageHomeIntro';
import PageHomeSkills from '../PageHomeSkills/PageHomeSkills';

describe('PageHome', () => {
  let wrapper;
  const props = {};
  const state = {
    app: AppStub,
    pageHome: PageHomeStub,
  };

  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('matches snapshot and renders correctly with data', () => {
    const { ReduxWrapper } = reduxHelper(PageHome, props, state);
    wrapper = mountRender(ReduxWrapper);
    expect(wrapper.find(PageWrapper).length).toEqual(1);
    expect(wrapper.find(PageHomeFeatured).length).toEqual(1);
    expect(wrapper.find(PageHomeIntro).length).toEqual(1);
    expect(wrapper.find(PageHomeSkills).length).toEqual(1);
    expect(wrapper.find(PageHomeBlog).length).toEqual(1);
    expect(wrapper.html).toMatchSnapshot();
  });

  it('renders correctly when there is no data', () => {
    const { ReduxWrapper } = reduxHelper(PageHome, props);
    wrapper = mountRender(ReduxWrapper);
    expect(wrapper.find(PageWrapper).length).toEqual(1);
    expect(wrapper.find(PageHomeFeatured).length).toEqual(0);
    expect(wrapper.find(PageHomeIntro).length).toEqual(0);
    expect(wrapper.find(PageHomeSkills).length).toEqual(0);
    expect(wrapper.find(PageHomeBlog).length).toEqual(0);
    expect(wrapper.html).toMatchSnapshot();
  });

  it('runs the fetchHomeData action if intro is not defined', () => {
    const newState = {
      ...state,
      pageHome: {
        ...state.pageHome,
        content: {
          ...state.pageHome.content,
          intro: null as any,
        },
      },
    };
    const { dispatch, ReduxWrapper } = reduxHelper(PageHome, props, newState);
    wrapper = mountRender(ReduxWrapper);
    expect(dispatch).toHaveBeenCalledTimes(1);
  });

  it("doesn't run the fetchHomeData action if the intro is defined", () => {
    const { dispatch, ReduxWrapper } = reduxHelper(PageHome, props, state);
    wrapper = mountRender(ReduxWrapper);
    expect(dispatch).toHaveBeenCalledTimes(0);
  });

  it('runs the correct actions when the Home data is successfully fetched', async (done) => {
    const { getActions, ReduxWrapper } = reduxHelper(PageHome, props);
    wrapper = mountRender(ReduxWrapper);

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: PageHomeStub,
        })
        .then(() => {
          const expected = [
            {
              type: 'pageHome/fetchHomeRequest',
            },
            {
              type: 'pageHome/fetchHomeSuccess',
              payload: PageHomeStub,
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });

  it('runs the correct actions when the Home data is not successfully fetched', async (done) => {
    const { getActions, ReduxWrapper } = reduxHelper(PageHome, props);
    wrapper = mountRender(ReduxWrapper);

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 500,
        })
        .then(() => {
          const expected = [
            {
              type: 'pageHome/fetchHomeRequest',
            },
            {
              type: 'pageHome/fetchHomeError',
            },
          ];
          const actual = getActions();
          expect(actual).toEqual(expected);
          done();
        });
    });
  });
});
