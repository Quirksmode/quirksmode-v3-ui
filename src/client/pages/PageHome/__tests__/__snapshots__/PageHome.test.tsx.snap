// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHome matches snapshot and renders correctly with data 1`] = `[Function]`;

exports[`PageHome renders correctly when there is no data 1`] = `[Function]`;
